### DISCUSSION TITLE
From which variables is the maven URL build up. RegexManagers

### DISCUSSION THREAD

**Discussion by timguy** [renovatebot/renovate#22707] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Hi,
as ivy.xml is not supported I wrote my own regexMatcher.

### What you tried so far.

I have tried to get the regexManagers running with an ivy.xml
So my ivy.xml is recognized and parts are captured in the capture group.

**Problem** my URL is build up like this as I can see in the logs:
`https://repo.maven.apache.org/maven2/commons-beanutils/undefined/maven-metadata.xml`
`"msg":"Content is not found for Maven url"`

**Expected**: 
`https://repo.maven.apache.org/maven2/commons-beanutils/commons-beanutils/maven-metadata.xml`

Instead of "undefined" I want the "common-beanutils" again.
Which variable or capture group I do have to fill that I don't get _undefined_?

This is how my ivy.xml looks a like:
`<dependency org="commons-beanutils" name="commons-beanutils" rev="1.9.2">`

An this is parte of my renovate.json:
```
  "regexManagers": [
    {
      "fileMatch": ["ivy\\.xml$"],
      "matchStrings": [
        "<dependency\\s+org=\"(?<depName>[^\"]+)\"\\s+name=\"(?<lookupName>[^\"]+)\"\\s+rev=\"(?<currentValue>[^\"]+)\""
      ],
      "datasourceTemplate": "maven",
      "versioningTemplate": "maven"
    }
  ],
```

I also tried `packageName `instead of `lookupName `but anyway the URL still contains the `undefined`. 

Thanks for help and hints.

The regex itself isn't the problem, all capture groups are hit (only double backslashs removed):
https://regex101.com/r/fQQx5z/1

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by viceice** [comment]
you need to combine org and name to a full package name via depNameTemplate and don't use `lookupName` as capture group, simply use any other group name which isn't known to renovate regex manager

**Comment by timguy** [comment]
thanks, this was helpful @viceice .
Now I am a bit closer:

```	
Is:
https://repo.maven.apache.org/maven2/commons-beanutils/commons-beanutils/undefined/maven-metadata.xml","msg":"Content is not found for Maven 
Expected:
https://repo.maven.apache.org/maven2/commons-beanutils/commons-beanutils/maven-metadata.xml
```

With this setting:
```
    "regexManagers": [
    {
      "fileMatch": ["ivy\\.xml$"],
      "matchStrings": [
        "<dependency\\s+org=\"(?<ivyOrg>[^\"]+)\"\\s+name=\"(?<ivyName>[^\"]+)\"\\s+rev=\"(?<currentValue>[^\"]+)\""
      ],
      "datasourceTemplate": "maven",
      "versioningTemplate": "maven",
      "depNameTemplate": "{{ivyOrg}}/{{ivyName}}"
    }
```

I get the repetition of commons-beanutils thanks to the depNameTemplate.
But there is still this undefined variable whatever it is attached here. 
If I would know which variable is used here in renovate I could use the ivyName for it directly .

**Reply by viceice** [reply, marked as answer]
depName needs to be `<org>:<name>`. 

that's the format which the maven datasource expects.

**Reply by timguy** [reply]
isch globs nicht. Läuft.
Thanks.

Here the solution for ivy.xml if others end up here:
```

    "regexManagers": [
    {
      "fileMatch": ["ivy\\.xml$"],
      "matchStrings": [
        "<dependency\\s+org=\"(?<ivyOrg>[^\"]+)\"\\s+name=\"(?<ivyName>[^\"]+)\"\\s+rev=\"(?<currentValue>[^\"]+)\""
      ],
      "datasourceTemplate": "maven",
      "versioningTemplate": "maven",
      "depNameTemplate": "{{ivyOrg}}:{{ivyName}}"
    }
  ],
```

**Comment by timguy** [comment]
Just for clarificatio @viceice.
Could I have found out this easily by myself?
Is there a documentation which parameter is expected by datasource (maven in this case)?
- I basically started from here: https://docs.renovatebot.com/modules/manager/regex/
- And documentation about datasource is also not more specific: https://docs.renovatebot.com/modules/datasource/maven/
I also read in maven about artefacts and groups and try to find a connection to the renovate world.

**Reply by viceice** [reply]
not sure, you can probably find it inside renovate debug logs for a pom file.
